bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -29,6 +31,75 @@ use crate::migrate::{
 use crate::vm::{MigrateSourceCommand, MigrateSourceResponse, VmController};
 
 /// Specifies which pages should be offered during a RAM transfer phase.
+///
+/// On Dirty Pages and the Discipline Thereof
+/// -----------------------------------------
+///
+/// In an ideal world, a migration would only ever transfer pages of the guest's
+/// address space which have actually been touched by the guest; we don't want
+/// to waste time sending a bunch of zero pages on the wire. RAM is offered to
+/// the migration destination in two phases: first, we transfer the majority of
+/// the guest's RAM prior to pausing the guest, and second, after pausing the
+/// guest, we transfer any pages which have been touched since when we performed
+/// the first transfer. This way, we perform most of the RAM transfer while the
+/// guest is still running, and only re-transfer pages which have been dirtied
+/// again while paused.
+///
+/// Transferring only dirty pages is made possible by bhyve's
+/// `VM_TRACK_DIRTY_PAGES` and `VM_NPT_OPERATION`` ioctls. The
+/// `VM_TRACK_DIRTY_PAGES` ioctl allows us to generate a bitmap of which pages
+/// have their dirty flags set, allowing us to offer only those pages when
+/// performing a RAM transfer. Because `VM_TRACK_DIRTY_PAGES`` also clears the
+/// dirty bits, if we call the ioctl when offering the initial pre-pause RAM
+/// transfer, and then again when performing the post-pause RAM transfer, the
+/// second ioctl call will see only the dirty bits that were set *since* when
+/// the initial transfer was performed, allowing us to offer only the pages
+/// which have been touched since we transferred most of the memory.
+///
+/// Sounds simple enough, right? Well, here's where things get interesting.
+/// Should a migration *fail* after transferring RAM, all the dirty bits on the
+/// guest's pages will have been cleared by the `VM_TRACK_DIRTY_PAGES` ioctl.
+/// This means that, for a naive implementation which just always transfers only
+/// the pages marked as dirty by the ioctl, a second or third migration attempt
+/// will not transfer any pages whose dirty bits were cleared by the first
+/// migration attempt and haven't been touched again since then. This is bad
+/// news! The guest has written to those pages, and, just because it hasn't
+/// touched them since the last migration attempt, it may still care about that
+/// memory, and attempt to read what it put there again in the future --- with
+/// unpleasant results, if we haven't transferred that memory.
+///
+/// There are two potential ways we can solve this, and --- as you're about to
```

Comment:
There is a third: scan each page and refrain from sending it if it's all zeroes. This may be a useful optimization in any event, as many "dirty" pages were probably just zeroed by the guest. Of course, it trades off CPU time for reduced latency and using less memory for maintaining the accumulated set of dirty pages, so that's got to be balanced against the requirement of not starving running guests, but considering that dirty pages can short-circuit the scan as soon as they find a non-zero byte (or u64) it may not be that bad.